onnxruntime/core/session/inference_session.cc

```diff
@@ -156,18 +156,14 @@ class InferenceSession::Impl {
 
     InitLogger(logging_manager);
 
-    // currently the threadpool is used by the parallel executor only and hence
-    // there is no point creating it when only sequential execution is enabled.
-    if (!session_options.enable_sequential_execution) {
-      int pool_size = session_options_.session_thread_pool_size == 0
+    // The threadpool is currently evolving.  We will always create a per session threadpool.
+    // Beyond this, we will create a global thread pool to share across sessions.
+    {
+      int pool_size = session_options_.session_thread_pool_size <= 0
                           ? std::thread::hardware_concurrency() / 2
                           : session_options_.session_thread_pool_size;
 
-#ifdef USE_EIGEN_THREADPOOL
-      thread_pool_ = std::make_unique<Eigen::NonBlockingThreadPool>(pool_size);
-#else
-      thread_pool_ = std::make_unique<TaskThreadPool>(pool_size);
-#endif
+      thread_pool_ = std::make_unique<onnxruntime::concurrency::ThreadPool>("SESSION", pool_size);
```

Comment:
Is there a reason for allocating the threadpool on the heap?

Reply:
Would this be an optional nature of heap allocation?

---
In reply to: [268015016](https://github.com/Microsoft/onnxruntime/pull/674#discussion_r268015016) [](ancestors = 268015016)

Reply:
The original code delayed initialization and allocated in this way.  I agree, it can be an expensive allocation.  Will look further.